apps/api/views.py

```diff
@@ -24,22 +26,30 @@ def get_experiments(request):
 
 @api_view(["POST"])
 @permission_classes([HasUserAPIKey])
-def update_participant_data(request):
+def update_participant_data(request, participant_id: UUID):
     """
     Upsert participant data for a specific experiment
     """
-    experiment_public_id = request.data["experiment_id"]
-    identifier = request.data["participant_id"]
-    new_data = request.data["details"]
-    experiment = get_object_or_404(Experiment, public_id=experiment_public_id, team=request.team)
-    participant = get_object_or_404(Participant, identifier=identifier, team=experiment.team)
-
-    try:
-        participant_data = experiment.participant_data.get(participant=participant)
-        participant_data.data = new_data
-        participant_data.save(update_fields=["data"])
-    except ParticipantData.DoesNotExist:
-        ParticipantData.objects.create(
-            participant=participant, data=new_data, content_object=experiment, team=experiment.team
-        )
+    experiment_data = request.data["data"]
+    experiment_ids = experiment_data.keys()
+    experiments = Experiment.objects.filter(public_id__in=experiment_ids, team=request.team)
+    experiment_map = {str(experiment.public_id): experiment for experiment in experiments}
+    participant = get_object_or_404(Participant, public_id=participant_id, team=request.team)
+
+    experiments_not_updated = []
+    for experiment_id, new_data in experiment_data.items():
+        if experiment_id not in experiment_map:
+            experiments_not_updated.append(experiment_id)
+            continue
+
+        experiment = experiment_map[experiment_id]
+
+        try:
+            participant_data = experiment.participant_data.get(participant=participant)
+            participant_data.data = new_data
+            participant_data.save(update_fields=["data"])
+        except ParticipantData.DoesNotExist:
+            ParticipantData.objects.create(
+                participant=participant, data=new_data, content_object=experiment, team=experiment.team
+            )
```

Comment:
I think you could replace this with `ParticipantData.objects.update_or_create`

Reply:
In normal circumstances that would have worked, but since the `data` field is encrypted in the DB, searching on that field doesn't play nice

Reply:
I don't quite follow. We don't need to search on data do we?:
```python
ParticipantData.objects.update_or_create(
    participant=participant, 
    content_object=experiment,
    defaults={"data": new_data},
    create_defaults={"team": experiment.team},
)```

Reply:
Sorry I meant the update query, but looks like your solution worked, since we don't have to actually search the data as well (except the usage of `create_defaults`, which was only introduced in django 5 I think. Django 4.2 doesn't have that, but it's fine)